### ISSUE TITLE
Define macro config based on other macros

### ISSUE THREAD

**Issue by mydea** [embroider-build/embroider#613]
It would be awesome if you could define a macro based on another macro. Concretely, this is what I'd love to do:

```js
const { isTesting } = require('@embroider/macros');

module.exports = {
  name: require('./package').name,

  options: {
    '@embroider/macros': {
      setOwnConfig: {
        isEnabled: !isTesting() && someOtherCheck,
      },
    },
  },
};
```

That would allow us to encapsulate such checks much nicer - now, I have to repeat this everywhere in our app like this:

```js
if (macroCondition(!isTesting() && isDevelopingApp() && getOwnConfig().someOtherCheck) {
  // ...
}
```

which kind of works but feels a bit more error prone than necessary.

**Comment by ef4** [comment]
Yeah, one idea for this area is that people can rely on the macros as a compile target and implement nicer APIs on top of them. The benefit of this strategy is that the macros themselves can remain stable and supported effectively forever, even as people experiment with different developer-facing APIs that compile down to the macros before publishing. It means you're free to bring your own API, but you'll always be compatible with everybody else who supports the macros.

That said, if we can come up with a really robust general-purpose proposal, we can build it right into the system.

It would be hard to make the above code example work correctly, because in development `isTesting()` can change its value even after the build has happened, so the expression `!isTesting() && someOtherCheck` would need to be re-evaluated at runtime.

An alternative sketch is to create a decorator that marks an entire function as becoming a macro itself, meaning that it would become a compile error if we can't statically determine its return value, and things like `macroCondition` would always be able to "see" the result.

```js
import { macro, isTesting, isDevelopingApp, getOwnConfig } from "@embroider/macros";

@macro
function isEnabled() {
  return !isTesting() && isDevelopingApp() && getOwnConfig().someOtherCheck;
}

if (macroCondition(isEnabled())) {
  // ...
}
```

The biggest challenge I see here is that I don't really want to defeat our ability to transpile modules in parallel, one by one. If you tried to import this `isEnabled` function from a different module, now we can't compile that module without first processing the one that defines `isEnabled`. 

So most likely we would need to be restrictive about where we let people define these macro functions. We need to be able to quickly find all of them, across an entirely project and all its dependencies. Really they belong as a peer to `setOwnConfig`, something like `defineMacros`, and they would probably need to be strings containing Javascript (because we wouldn't want them to look like functions that actually run inside the Node configuration code, because they don't).

**Comment by mydea** [comment]
Yeah, that sounds reasonable! Being able to define such (simple) macros ourselves would definitely be very nice :+1:
